strapi/strapi
# Fix entity-service types for relations' input

### What does it do?

- Add support for relation reordering APIs (only for x(to)Many relations)
	- set
	- connect (including positional arguments `before`, `after`, `start` and `end`)
	- disconnect
- Add support for both shorthand and longhand notations (`ID | { id: ID }`) for all relations

### Why is it needed?

Fix broken types introduced by the Entity-Service x TypeScript PR

### How to test it?

In a Strapi project with types, write a create or update query and mutate relation properties.
- For `x(to)Many` relations, it should propose both the reordering APIs (`set`, `connect` and `disconnect`) and an array of either `ID`s or `{ id: Id }`
- For `x(to)One` relations, it should propose either `ID` or `{ id: ID }`

### Examples

Note: Using the `admin::role` generated type for this example

![image](https://github.com/strapi/strapi/assets/25851739/b8a33068-a252-466a-ab90-a9847f7c274d)

```ts
const { update } = strapi.entityService;
const uid = 'admin::user' as const;

// xMany (valid)
await update(uid, 1, { data: { roles: null } });
await update(uid, 1, { data: { roles: [] } });
await update(uid, 1, { data: { roles: [2, '3'] } });
await update(uid, 1, { data: { roles: [{ id: 2 }, { id: '3' }] } });
await update(uid, 1, { data: { roles: { set: null } } });
await update(uid, 1, { data: { roles: { set: [] } } });
await update(uid, 1, { data: { roles: { set: [2, 3] } } });
await update(uid, 1, { data: { roles: { set: [{ id: 2 }, { id: 3 }] } } });
await update(uid, 1, { data: { roles: { connect: [1, 2] } } });
await update(uid, 1, { data: { roles: { connect: [{ id: 1 }, { id: 2 }] } } });
await update(uid, 1, { data: { roles: { disconnect: [1, 2] } } });
await update(uid, 1, { data: { roles: { disconnect: [{ id: 1 }, { id: 2 }] } } });
await update(uid, 1, {
  data: { roles: { connect: [1, 2], disconnect: [{ id: 1 }, { id: 2 }] } },
});
// xMany (errors)
await update(uid, 1, { data: { roles: 2 } }); // not assignable to number
await update(uid, 1, { data: { roles: [2, { id: 2 }] } }); // number not assignable to { id: ID }

// xOne (valid)
await update(uid, 1, { data: { createdBy: null } });
await update(uid, 1, { data: { createdBy: 1 } });
await update(uid, 1, { data: { createdBy: '2' } });
// xOne (errors)
await update(uid, 1, { data: { createdBy: {} } }); // not assignable to {}
await update(uid, 1, { data: { createdBy: [] } }); // not assignable to []
```

### Related issue(s)/PR(s)

fix #18106
fix #18317
